File: EV-877/MARC.Everest/DataTypes/Interfaces/IEncapsulatedData.cs
/* 
 * Copyright 2008-2012 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.

 * 
 * User: Justin Fyfe
 * Date: 01-09-2009
 */
using System;
using System.Collections.Generic;
using System.Text;
using MARC.Everest.Attributes;
using System.ComponentModel;

namespace MARC.Everest.DataTypes.Interfaces
{

    /// <summary>
    /// Supported data compression algorithms
    /// </summary>
    /// <remarks>
    /// Only the two formats supported by the system are listed
    /// </remarks>
    [Structure(Name = "CompressionAlgorithm", CodeSystem = "2.16.840.1.113883.5.1009", StructureType = StructureAttribute.StructureAttributeType.ConceptDomain)]
    public enum EncapsulatedDataCompression
    {
        /// <summary>
        /// Deflate
        /// </summary>
        DF,
        /// <summary>
        /// GZIP
        /// </summary>
        GZ,
        /// <summary>
        /// ZLib
        /// </summary>
        ZL,
        /// <summary>
        /// Compress
        /// </summary>
        Z,
        /// <summary>
        /// BZIP
        /// </summary>
        BZ,
        /// <summary>
        /// 7z Compression
        /// </summary>
        Z7
    }

    /// <summary>
    /// Representation of the encapsulated data
    /// </summary>
    public enum EncapsulatedDataRepresentation
    {
        /// <summary>
        /// Data should be represented in plain text
        /// </summary>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "TXT")]
        [Enumeration(Value = "TXT")]
        TXT,
        /// <summary>
        /// Data should be base64 encoded
        /// </summary>
        [Enumeration(Value = "B64")]
        B64,
        /// <summary>
        /// Data is xml encoded
        /// </summary>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "XML")]
        [Enumeration(Value = "TXT")]
        XML
    }

    /// <summary>
    /// Representation of how an ED should be checksummed
    /// </summary>
    [Structure(Name = "IntegrityCheckAlgorithm", CodeSystem = "2.16.840.1.113883.5.1010", StructureType = StructureAttribute.StructureAttributeType.ConceptDomain)]
    public enum EncapsulatedDataIntegrityAlgorithm
    {
        /// <summary>
        /// Using a SHA1 hash
        /// </summary>
        [Enumeration(Value = "SHA-1")]
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SHA")]
        SHA1,
        /// <summary>
        /// Using a SHA256 hash
        /// </summary>
        [Enumeration(Value = "SHA-256")]
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Naming", "CA1709:IdentifiersShouldBeCasedCorrectly", MessageId = "SHA")]
        SHA256
    }

    /// <summary>
    /// Identifies that a class implements (or is compatible with) an IEncapsulatedData 
    /// </summary>
    [EditorBrowsable(EditorBrowsableState.Never)]
    public interface IEncapsulatedData
    {
        /// <summary>
        /// Represents the data contained within the encapsulated data
        /// </summary>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1819:PropertiesShouldNotReturnArrays")]
        byte[] Data { get; set; }

        /// <summary>
        /// Get or set the desired representation
        /// </summary>
        EncapsulatedDataRepresentation Representation { get; set; }

        /// <summary>
        /// Get or set the language
        /// </summary>
        string Language { get; set; }

        /// <summary>
        /// Get or set the media type
        /// </summary>
        string MediaType { get; set; }

        /// <summary>
        /// Get the integrity check
        /// </summary>
        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Performance", "CA1819:PropertiesShouldNotReturnArrays")]
        byte[] IntegrityCheck { get; }

        /// <summary>
        /// Get or set the integrity check algorithm  to use
        /// </summary>
        EncapsulatedDataIntegrityAlgorithm? IntegrityCheckAlgorithm { get; set; }

        /// <summary>
        /// Get or set a thumbnail represetnation of this ED
        /// </summary>
        IEncapsulatedData Thumbnail { get;}

        /// <summary>
        /// Get or set the reference to the data
        /// </summary>
        ITelecommunicationAddress Reference { get;}

        /// <summary>
        /// Get or set the compression algorithm used for the data
        /// </summary>
        EncapsulatedDataCompression? Compression { get; }

    }
}

File: jEverest/src/ca/marc/everest/datatypes/interfaces/IEncapsulatedData.java
/* 
 * Copyright 2008-2012 Mohawk College of Applied Arts and Technology
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you 
 * may not use this file except in compliance with the License. You may 
 * obtain a copy of the License at 
 * 
 * http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * User: Justin Fyfe
 * Date: 10-02-2012
 */
package ca.marc.everest.datatypes.interfaces;

import ca.marc.everest.datatypes.EncapsulatedDataCompression;
import ca.marc.everest.datatypes.EncapsulatedDataIntegrityAlgorithm;
import ca.marc.everest.datatypes.EncapsulatedDataRepresentation;

/**
 * Represents a type that encapsulates data in a message
 */
public interface IEncapsulatedData {

	/**
	 * Gets the data within the encapsulated data object
	 */
	byte[] getData();
	/**
	 * Sets the data carried in the encapsulated data object
	 */
	void setData(byte[] value);

	/**
	 * Gets the representation that the encapsulated data should carry on the wire
	 */
	EncapsulatedDataRepresentation getRepresentation();
	/**
	 * Sets a value instructing the formatter how to represent the data carried in 
	 * the encapsulated data
	 */
	void setRepresentation(EncapsulatedDataRepresentation value);

	/**
	 * Gets the language code representing the language of the content within the encapsulated
	 * data instance
	 */
	String getLanguage();
	/**
	 * Sets the language code representing the language of the content within the encapsulated 
	 * data instance
	 */
	void setLanguage(String value);

	/**
	 * Gets the content-type of the content carried within the encapsulated data
	 */
	String getMediaType();
	/**
	 * Sets the content-type of the content carried within the encapsulated data instance
	 */
	void setMediaType(String value);
	
	/**
	 * Gets a series of byte representing the integrity check of the data carried in the encapsulated
	 * data instance.
	 */
	byte[] getIntegrityCheck();

	/**
	 * Gets a value representing the algorithm used to calculate the integrity check
	 */
	EncapsulatedDataIntegrityAlgorithm getIntegrityCheckAlgorithm();
	/**
	 * Sets a value representing the algorithm used to calculate the integrity check property
	 * of the encapsulated data instance
	 */
	void setIntegrityCheckAlgorithm(EncapsulatedDataIntegrityAlgorithm value);

	/**
	 * Gets an encapsulated data instance which is a thumbnail representation of the encapsulated data
	 * instance
	 */
	IEncapsulatedData getThumbnail();

	/**
	 * Gets a telecommunications address which points to the contents of the encapsulated data instance
	 */
	ITelecommunicationsAddress getReference();
	
	/**
	 * Gets a value representing the compression algorithm used to compress the data encapsulated in this instance
	 */
	EncapsulatedDataCompression getCompression();
}
